source/l4_testing/wheel_speeds/wheel_speeds.c

```diff
@@ -0,0 +1,70 @@
+/**
+ * @file wheel_speeds.c
+ * @author Luke Oxley (lcoxley@purdue.edu)
+ * @brief   Use pwm input capture to measure left and right wheel speeds
+ * @version 0.1
+ * @date 2022-01-21
+ * 
+ * @copyright Copyright (c) 2021
+ * 
+ */
+
+#include "wheel_speeds.h"
+#include "can_parse.h"
+
+WheelSpeeds_t wheel_speeds;
+
+uint32_t left_ccr_msb = 0;
+uint32_t left_ccr_lsb = 0;
+uint32_t right_ccr = 0;
+
+extern q_handle_t q_tx_can;
+extern uint32_t APB2ClockRateHz;
+
+void wheelSpeedsInit()
+{
+    // enable interrupts
+    // TIM1 (left)
+    TIM1->DIER |= TIM_DIER_CC1IE | TIM_DIER_UIE;
+    NVIC_SetPriority(TIM1_UP_TIM16_IRQn, 1);
+    NVIC_SetPriority(TIM1_CC_IRQn, 2);
+    NVIC_EnableIRQ(TIM1_UP_TIM16_IRQn);
+    NVIC_EnableIRQ(TIM1_CC_IRQn);
+    PHAL_startTIM(TIM1);
+
+    // TIM2 (right)
+    // TIM2->DIER |= TIM_DIER_CC1IE;
+    // NVIC_EnableIRQ(TIM2_IRQn);
+    // PHAL_startTIM(TIM2);
+}
+
+void wheelSpeedsPeriodic()
+{
+    wheel_speeds.left.freq_hz = APB2ClockRateHz / TIM_PRESC / ((left_ccr_msb << 16) | left_ccr_lsb);
```

Comment:
Might want to combine these values inside the IRQ or something like this:
```suggestion
do {
    left_ccr_comb = (left_ccr_msb << 16) | left_ccr_lsb;
} while(left_ccr_comb != (left_ccr_msb << 16) | left_ccr_lsb)
```
You could have an TIM1_CC_IRQ in the middle of combining the MSB and LSB that would lead to an incorrect combination of your msb and lsb.

Reply:
What is the purpose of the do while?